Issue #36 - Adding VPC Endpoint Support

davidtkaczyk69 opened the pull request:
The direct use of specific VPC Endpoints or standard regional KMS
endpoints was not supported by starter as is.  New configuration
properties will allow service endpoint and an optional signing region to
be specified.

harti2006 commented on src/main/java/de/zalando/spring/cloud/config/aws/kms/KmsProperties.java:
```diff
@@ -23,6 +23,46 @@
+     */
+    private Endpoint endpoint;
+    
+    @ConfigurationProperties("endpoint")
```
This annotation can be omitted. See docs about [nested properties](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#configuration-metadata-nested-properties).

harti2006 commented on src/main/java/de/zalando/spring/cloud/config/aws/kms/KmsEncryptionConfiguration.java:
```diff
@@ -52,20 +54,42 @@ KmsTextEncryptor kmsTextEncryptor() {
+    }
 
     @Configuration
+    @ConditionalOnProperty(name="aws.kms.region", matchIfMissing=true)
```
This doesn't work for me. Spring never picks the `KmsConfigurationEndpoint` class because the region condition is alway true. I would suggest to not use `ConditionalOnProperty` here at all and just handle the business logic the `kms()` bean method, like:

```
if (serviceEndpoint present) {
  builder.withEndpointConfig(serviceEndpoint)
}

davidtkaczyk69 replied:
Fixed.

harti2006 commented:
Alright, I will review and test again.

harti2006 commented:
It works perfectly fine on my machine. Still I'm thinking about how to unit/integration test this new change. Maybe some test with Wiremock could be done here. Do you have other ideas, @davidtkaczyk69 ?

davidtkaczyk69 commented:
With regard to unit testing, I was wondering that myself.  Let me see if I can figure something similar to what is being done in the bootstrap-noEncryption.yml setup.

davidtkaczyk69 commented:
Added junit tests per your suggestion and have pushed to my branch.  Uses Reflection to validate that AWSKMS bean was populated properly.  Not a big fan of introducing reflection here, but it does prove that the underlying bean was injected with bootstrap property data correctly.  Modified existing test to also verify that region property was still valid as well.  This bean is mocked in the "encryption" profile junit tests, so I didn't bother to modify those as it wouldn't have proved anything.  Please let me know what you think.  Open to other approaches/ideas.

davidtkaczyk69 commented:
Follow up junit commit changed pushed to remove unused import.

harti2006 commented:
Awesome. Thank you very much for your contribution.

harti2006 commented:
:+1:

ChristianLohmann commented:
:+1:

harti2006 commented:
@davidtkaczyk69 it's released in version 2.0 http://search.maven.org/#artifactdetails%7Corg.zalando%7Cspring-cloud-config-aws-kms%7C2.0%7Cjar
